IMG/man/select.1
.TH SELECT 1 "MIDI"
.SH NAME
select \- select specified portions of MIDI data
.SH SYNOPSIS
.B select
[\fB\-c\fP#[\fB\-\fP#]]
[\fB\-k\fP#[\fB\-\fP#]]
[\fB\-v\fP#[\fB\-\fP#]]
[\fB\-A\fP]
[\fB\-C\fP#[\fB\-\fP#]]
[\fB\-P\fP#[\fB\-\fP#]]
[\fB\-B\fP]
[\fB\-S\fP#[\fB\-\fP#]]
[\fB\-only\fP]
[\fB\-allbut\fP]
<file1 >file2
.SH DESCRIPTION
.I Select
reads MIDI command streams
and passes through those portions that meet the specified selection criteria.
Selection can be based on any combination of criteria: channel,
key (note), key velocity, and status commands.
.PP
Unless the \fB\-only\fP argument is used, the defaults are to let everything
through (equivalent to:
\fB\-c1\-16 \-k0\-127 \-v0\-127 \-C0\-127 \-P1\-128 \-S0\-15\fP).
When an argument is specified, the corresponding default argument is ignored,
thus \fB\-c3\fP overrides \fB\-c1\-16\fP in the default.
When \fB\-only\fP is used, the default is to let nothing through
(equivalent to: \fB\-c \-k \-v \-C \-P \-S\fP).
.PP
The \fB\-allbut\fP option complements the selection criteria, however it does
not complement the effect of \fB\-only\fP; (see the examples).
.PP
The \fB\-c\fP argument specifies one or more channel numbers to be selected.
\fB\-c1\fP selects the first MIDI channel, \fB\-c1\-16\fP selects all MIDI
channels (the default).  To select disjoint ranges of channels use more than
one argument: e.g. \fB\-c2 \-c4\fP selects the second and fourth MIDI channels.
.PP
The \fB\-k\fP argument specifies one or more notes (keys) to be selected.
\fB\-k60\fP selects C4 (or cymbal crash on an RX11/15), \fB\-k0\-127\fP selects
all the notes (the default).  To select disjoint ranges of notes use more than
one argument as described above.
.PP
The \fB\-v\fP argument specifies one or more key velocities to be selected.
\fB\-v64\fP selects only notes played with a velocity of 64, \fB\-v1\-127\fP
selects all key velocities (the default).
Note that key velocity 0 is \fIalways\fP selected since many MIDI devices
use a key velocity of 0 to indicate key-off.
To remove the unneeded key-off commands after filtering out some key-on
commands, use mpuclean(1).
To select disjoint ranges of key velocities use more than one argument
as described above.
.PP
The \fB\-C\fP argument specifies a control change command to be selected.
\fB\-C1\fP selects modulation wheel commands;
\fB\-C2\fP selects breath control commands;
\fB\-C3\fP selects foot control commands;
\fB\-C7\fP selects volume commands;
\fB\-C\fP deselects all control change commands;
\fB\-C1\-4 \-C6\-128\fP selects all but portamento commands;
et cetera.
.PP
The \fB\-P\fP argument specifies a program change command to be selected.
\fB\-P1\fP selects program 1 change commands;
\fB\-P\fP deselects all program change commands;
\fB\-P1\-5 \-P7\-128\fP selects all but program 6 change commands;
et cetera.
.PP
The \fB\-A\fP argument deselects polyphonic after-touch messages.
.PP
The \fB\-B\fP argument deselects pitch-bend messages.
.PP
The \fB\-S\fP argument specifies a system message command to be selected.
\fB\-S0\fP selects system exclusive messages;
\fB\-S2\fP selects song position pointer messages;
\fB\-S3\fP selects song select messages;
\fB\-S6\fP selects tune request messages;
\fB\-S7\fP selects end-of-SystemExclusive messages;
\fB\-S8\fP selects timing clock messages;
\fB\-S9\fP selects 0xf9 messages (used by the Roland MPW-401 and
MPU-401-compatible devices for ``timing clock with measure end'');
\fB\-S10\fP selects start messages;
\fB\-S11\fP selects continue messages;
\fB\-S12\fP selects stop messages;
\fB\-S14\fP selects active sensing messages;
\fB\-S15\fP selects system reset messages;
To select several kinds of system messages use several \fB\-S\fP arguments.
.PP
All MIDI data are read from the standard input and written on the
standard output.
.PP
.SH EXAMPLES
.Cs
select \-c1\-2 <bach >rebach
.Ce
This command will copy all MIDI commands from bach that refer to channels
1 or 2 (and any commands that don't refer to channels at all).
.Cs
select \-k0\-63 <bach >lobach
select \-allbut \-k0\-63 <bach >hibach
.Ce
These commands will split "bach" into "lobach" and "hibach" containing the
left and right halves of the keyboard respectively.
Note that both files created will contain whatever miscellaneous program changes
or other, non-key-related MIDI commands appeared in the original file.
.Cs
select \-v10\-127 <bach >bebach
.Ce
This will filter out all the very quiet notes.
.Cs
select \-allbut \-only \-c1
select \-only \-c2\-16
.Ce
These two commands do the same thing.
.SH SEE ALSO
chmap(1), filter(1), kmap(1), mpuclean(1)
.SH AUTHOR
Peter Langston, Bell Communications Research
(bellcore!psl)


IMG/select.c
/*
**	SELECT -- Select specified parts of MIDI input
**	psl 11/85
*/
#include <stdlib.h>
#include	<stdio.h>
#include	"midi.h"

#define	MAXCHAN	MIDI_MAX_CHANS
#define	MAXKEY	128
#define	MAXVEL	128
#define	MAXCNT	128
#define	MAXPRG	128
#define	MAXSYS	16


void argval(char *arg, int map[], int max, int offset);


char	*Progname;
	/* message types */
int	Nall	= 1;			/* note-on/note-off */
int	Aall	= 1;			/* after touch */
int	Call	= 1, Cmap[MAXCNT];	/* control changes */
int	Pall	= 1, Pmap[MAXPRG];	/* program changes */
int	Ball	= 1;			/* pitch bend */
int	Sall	= 1, Smap[MAXSYS];	/* sys & misc */
	/* secondary criteria (requires Nall) */
int	Kall	= 1, Kmap[MAXKEY];	/* key numbers */
int	Vall	= 1, Vmap[MAXVEL];	/* key velocities */
	/* orthogonal criterion */
int	Chall	= 1, Chmap[MAXCHAN];	/* channels */

syntax(f, s)
char	*f, *s;
{
	fprintf(stderr, f, s);
	fprintf(stderr, "Usage: %s [options] <old >new\n", Progname);
	fprintf(stderr, "Options are:\n");
	fprintf(stderr, "-c[#[-#]]\t-c3 for channel 3, -c for no channels\n");
	fprintf(stderr, "-k[#[-#]]\t-k61 for C#4, -k for no key events\n");
	fprintf(stderr, "-v[#[-#]]\t-v64 for key vel 64, -v0 for key-off\n");
	fprintf(stderr, "-C[#[-#]]\t-C7 for vol, -C for no control changes\n");
	fprintf(stderr, "-P[#[-#]]\t-P7 for prog 7, -P for no prog changes\n");
	fprintf(stderr, "-A\t\t-A for no after-touch\n");
	fprintf(stderr, "-B\t\t-B for no pitch-bend\n");
	fprintf(stderr, "-S[#[-#]]\t-S0 for sys excl, -S for no sys or realtime\n");
	fprintf(stderr, "-allbut\t\tto complement criteria (except -only)\n");
	fprintf(stderr, "-only\t\tto change defaults (see below)\n");
	fprintf(stderr, "Defaults without '-only':\n\t");
	fprintf(stderr, "-c1-16  -k0-127  -v0-127  -C0-127  -P1-128  -S0-15\n");
	fprintf(stderr, "Defaults with '-only':\n\t");
	fprintf(stderr, "-c  -k  -v  -C  -P  -S\n");
	exit(2);
}

main(argc, argv)
char	*argv[];
{
	register int status, type, i, allbut;
	long now;
	MCMD *mp;

	Progname = argv[0];
	for (i = 1; i < argc; i++) {		/* pre-pass for only */
	    if (argv[i][0] == '-') {
		switch (argv[i][1]) {
		case 'o':		/* ONLY */
		    Nall = Call = Pall = Aall = Ball = Sall = 0;
		    break;
		}
	    }
	}
	allbut = 0;
	for (i = 1; i < argc; i++) {		/* main pass */
	    if (argv[i][0] == '-') {
		switch (argv[i][1]) {
		case 'A':		/* after-touch */
		    Aall = 0;
		    break;
		case 'a':		/* ALLBUT */
		    allbut ^= 1;
		    break;
		case 'B':		/* pitch-bend */
		    Ball = 0;
		    break;
		case 'C':		/* control change(s) */
		    argval(&argv[i][2], Cmap, MAXCNT, 0);
		    Call = 0;
		    break;
		case 'c':		/* channel number(s) */
		    argval(&argv[i][2], Chmap, MAXCHAN, -1);
		    Chall = 0;
		    break;
		case 'k':		/* key number(s) */
		    argval(&argv[i][2], Kmap, MAXKEY, 0);
		    Kall = 0;
		    Nall = 1;
		    break;
		case 'o':		/* ONLY */
		    break;
		case 'P':		/* program change(s) */
		    argval(&argv[i][2], Pmap, MAXPRG, -1);
		    Pall = 0;
		    break;
		case 'S':		/* system & misc */
		    argval(&argv[i][2], Smap, MAXSYS, 0);
		    Sall = 0;
		    break;
		case 'v':		/* keyvel number(s) */
		    argval(&argv[i][2], Vmap, MAXVEL, 0);
		    Vall = 0;
		    Nall = 1;
		    break;
		default:
		    syntax("Unrecognized option arg: %s\n", argv[i]);
		}
	    } else
		syntax("Unrecognized arg: %s\n", argv[i]);
	}
	if (allbut) {
	    for (i = MAXCHAN; --i >= 0; Chmap[i] ^= 1);
	    for (i = MAXKEY; --i >= 0; Kmap[i] ^= 1);
	    for (i = MAXVEL; --i >= 0; Vmap[i] ^= 1);
	    for (i = MAXCNT; --i >= 0; Cmap[i] ^= 1);
	    for (i = MAXPRG; --i >= 0; Pmap[i] ^= 1);
	    for (i = MAXSYS; --i >= 0; Smap[i] ^= 1);
	}
	now = 0L;
	while (mp = getmcmd(stdin, now)) {
	    now = mp->when;
	    status = mp->cmd[0];
	    type = (status & M_CMD_MASK);
	    if (type == 0xF0) {			/* these have no channel */
		if (Sall || Smap[status & 0x0F])
		    putmcmd(stdout, mp);
		continue;
	    }
	    if (!Chall && !Chmap[status & M_CHAN_MASK])
		continue;
	    if (type == CH_KEY_OFF) {
		if (Nall
		 && (Kall || Kmap[mp->cmd[1]])
		 && (Vall || Vmap[0]))
		    putmcmd(stdout, mp);
	    } else if (type == CH_KEY_ON) {
		if (Nall
		 && (Kall || Kmap[mp->cmd[1]])
		 && (Vall || Vmap[mp->cmd[2]]))
		    putmcmd(stdout, mp);
	    } else if (type == CH_POLY_KPRS) {
		if (Aall)
		    putmcmd(stdout, mp);
	    } else if (type == CH_CTL) {
		if (Call || Cmap[mp->cmd[1]])
		    putmcmd(stdout, mp);
	    } else if (type == CH_PRG) {
		if (Pall || Pmap[mp->cmd[1]])
		    putmcmd(stdout, mp);
	    } else if (type == CH_PRESSURE) {
		if (Aall)
		    putmcmd(stdout, mp);
	    } else if (type == CH_P_BEND) {
		if (Ball)
		    putmcmd(stdout, mp);
	    } else {
		fprintf(stderr, "Weird MIDI status : %02x\n", status);
		putmcmd(stdout, mp);
	    }
	}
	exit(0);
}

void argval(arg, map, max, offset)
char	*arg;
int	map[];
{
	register char *cp;
	register int i, lo, hi;

	if (*arg == '\0')
	    return;
	lo = hi = atoi(arg) + offset;
	for (cp = arg; '0' <= *cp && *cp <= '9'; cp++);
	if (*cp == ':' || *cp == '-') {
	    if ((i = atoi(++cp) + offset) < lo)
		hi = max - 1;
	    else
		hi = i;
        }
	if (lo < 0 || hi >= max)
	    syntax("Value in '%s' out of range\n", arg);
	for (i = lo; i <= hi; map[i++] = 1);
}
